### ISSUE TITLE
race condition in services.ec2

### ISSUE THREAD

**Issue by jantman** [jantman/awslimitchecker#229]
This traceback is from a custom wrapper around awslimitchecker, that catches rate limiting/Throttling exceptions and raises other ones (but loses the traceback):

```
...
INFO:awslimitchecker.connectable:Connected to ec2 in region us-east-1
INFO:awslimitchecker.connectable:Connected to ec2 (resource) in region us-east-1
INFO:awslimitchecker.services.ec2:Querying EC2 DescribeAccountAttributes for limits
Checking AWS resource usage; WARNING threshold 80% of limit, CRITICAL threshold 99% of limit
Traceback (most recent call last):
  File "./check_aws_limits.py", line 410, in <module>
    checker.run(error_on_warning=args.error_on_warning, verbose=args.verbose)
  File "./check_aws_limits.py", line 318, in run
    warnings, criticals = self.run_for_regions(regions, verbose=verbose)
  File "./check_aws_limits.py", line 299, in run_for_regions
    w, c = self.run_for_region(region=region, verbose=verbose)
  File "./check_aws_limits.py", line 270, in run_for_region
    return self.check_limits(region, verbose=verbose)
  File "./check_aws_limits.py", line 206, in check_limits
    raise ex
botocore.exceptions.ClientError: An error occurred (InvalidInstanceID.NotFound) when calling the DescribeInstances operation: The instance ID 'i-2049342e' does not exist
```

It appears that the issue is in [_instance_usage()](https://github.com/jantman/awslimitchecker/blob/7bb5b788c12acbc7cf4d103013dccf39693a56a7/awslimitchecker/services/ec2.py#L247) which iterates through the instances as a Resource, and then checks properties on them. From what I can tell without a traceback, in rare cases instances can either be totally gone or not query-able yet when they're iterated. We should handle this gracefully.

**Comment by jantman** [comment]
Pushing this back to 0.8.0; need to get the traceback from that wrapper if/when it happens again...

**Comment by jantman** [comment]
Ok, got the traceback for this happening with [0.6.0](https://github.com/jantman/awslimitchecker/tree/0.6.0):

```
INFO:awslimitchecker.services.ec2:Querying EC2 DescribeAccountAttributes for limits
ERROR:__main__:Got unexpected (non-throttling) exception
Traceback (most recent call last):
  File "./check_aws_limits.py", line 197, in check_limits
    checker.check_thresholds()
  File "/var/jenkins_home/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/local/lib/python2.7/site-packages/awslimitchecker/checker.py", line 425, in check_thresholds
    tmp = cls.check_thresholds()
  File "/var/jenkins_home/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/local/lib/python2.7/site-packages/awslimitchecker/services/base.py", line 281, in check_thresholds
    self.find_usage()
  File "/var/jenkins_home/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/local/lib/python2.7/site-packages/awslimitchecker/services/ec2.py", line 69, in find_usage
    self._find_usage_instances()
  File "/var/jenkins_home/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/local/lib/python2.7/site-packages/awslimitchecker/services/ec2.py", line 81, in _find_usage_instances
    inst_usage = self._instance_usage()
  File "/var/jenkins_home/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/local/lib/python2.7/site-packages/awslimitchecker/services/ec2.py", line 247, in _instance_usage
    for inst in self.resource_conn.instances.all():
  File "/var/jenkins_home/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/local/lib/python2.7/site-packages/boto3/resources/collection.py", line 83, in __iter__
    for page in self.pages():
  File "/var/jenkins_home/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/local/lib/python2.7/site-packages/boto3/resources/collection.py", line 166, in pages
    for page in pages:
  File "/var/jenkins_home/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/local/lib/python2.7/site-packages/botocore/paginate.py", line 102, in __iter__
    response = self._make_request(current_kwargs)
  File "/var/jenkins_home/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/local/lib/python2.7/site-packages/botocore/paginate.py", line 174, in _make_request
    return self._method(**current_kwargs)
  File "/var/jenkins_home/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/local/lib/python2.7/site-packages/botocore/client.py", line 251, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/var/jenkins_home/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/local/lib/python2.7/site-packages/botocore/client.py", line 537, in _make_api_call
    raise ClientError(parsed_response, operation_name)
ClientError: An error occurred (InvalidInstanceID.NotFound) when calling the DescribeInstances operation: The instance IDs 'i-0a1a5cb3fb472c5ea, i-02cc14dfc6cba644e' do not exist
Checking AWS resource usage; WARNING threshold 80% of limit, CRITICAL threshold 99% of limit
Traceback (most recent call last):
  File "./check_aws_limits.py", line 411, in <module>
    checker.run(error_on_warning=args.error_on_warning, verbose=args.verbose)
  File "./check_aws_limits.py", line 319, in run
    warnings, criticals = self.run_for_regions(regions, verbose=verbose)
  File "./check_aws_limits.py", line 300, in run_for_regions
    w, c = self.run_for_region(region=region, verbose=verbose)
  File "./check_aws_limits.py", line 271, in run_for_region
    return self.check_limits(region, verbose=verbose)
  File "./check_aws_limits.py", line 197, in check_limits
    checker.check_thresholds()
  File "/var/jenkins_home/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/local/lib/python2.7/site-packages/awslimitchecker/checker.py", line 425, in check_thresholds
    tmp = cls.check_thresholds()
  File "/var/jenkins_home/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/local/lib/python2.7/site-packages/awslimitchecker/services/base.py", line 281, in check_thresholds
    self.find_usage()
  File "/var/jenkins_home/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/local/lib/python2.7/site-packages/awslimitchecker/services/ec2.py", line 69, in find_usage
    self._find_usage_instances()
  File "/var/jenkins_home/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/local/lib/python2.7/site-packages/awslimitchecker/services/ec2.py", line 81, in _find_usage_instances
    inst_usage = self._instance_usage()
  File "/var/jenkins_home/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/local/lib/python2.7/site-packages/awslimitchecker/services/ec2.py", line 247, in _instance_usage
    for inst in self.resource_conn.instances.all():
  File "/var/jenkins_home/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/local/lib/python2.7/site-packages/boto3/resources/collection.py", line 83, in __iter__
    for page in self.pages():
  File "/var/jenkins_home/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/local/lib/python2.7/site-packages/boto3/resources/collection.py", line 166, in pages
    for page in pages:
  File "/var/jenkins_home/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/local/lib/python2.7/site-packages/botocore/paginate.py", line 102, in __iter__
    response = self._make_request(current_kwargs)
  File "/var/jenkins_home/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/local/lib/python2.7/site-packages/botocore/paginate.py", line 174, in _make_request
    return self._method(**current_kwargs)
  File "/var/jenkins_home/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/local/lib/python2.7/site-packages/botocore/client.py", line 251, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/var/jenkins_home/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/local/lib/python2.7/site-packages/botocore/client.py", line 537, in _make_api_call
    raise ClientError(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (InvalidInstanceID.NotFound) when calling the DescribeInstances operation: The instance IDs 'i-0a1a5cb3fb472c5ea, i-02cc14dfc6cba644e' do not exist
```

Unfortunately, looking at [the code](https://github.com/jantman/awslimitchecker/blob/0.6.0/awslimitchecker/services/ec2.py#L247) and the traceback itself, this seems to be a problem in boto3/botocore itself, where pagination of ``conn.instances.all()`` fails if an instance disappears during the API call.

This run was using boto3 1.4.3 and botocore 1.4.93.

I've opened https://github.com/boto/boto3/issues/977 for this.

**Comment by jantman** [comment]
I ran some code to try and reproduce this, for almost 3 days. No luck. Closing both boto/boto3#977 and this issue, as it seems like this was fixed in botocore somewhere between 1.4.93 and 1.5.8. Will reopen if I can reproduce in the future...
